### DISCUSSION TITLE
.gitIgnore challenge - adds 6 seconds to each search

### DISCUSSION THREAD

**Discussion by csmwww** [dnGrep/dnGrep#1210] [Q&A] [answered]
Thank you for a very useful tool in almost all cases.

However, my common usage is to search code in a large project with many subdirectories and files (loaded with generated artifacts).

For this usage .gitIgnore is a perfect optimization as an alternative to "ignore directories" I've used in other tools, and yet dnGrep then rescans for .gitIgnore on every search adding a 6 second stall at the head of each search despite the fact .gitIgnore files are very rarely altered. Six seconds sound small, but my frequent use makes the impact very significant.

Is there any chance there is a setting I am missing that might allow dnGrep to cache .gitIngnore info, perhaps for the life of "Folder" selection, or some other creative interval (or manual)?

**Comment by doug24** [comment]
This is an interesting question.

Can you describe the nature of your repo where are seeing a 6 second difference?
Are you searching a single repo from the root of the repo, or multiple repos with multiple .gitignore files (where dnGrep must search down the directory tree to find .gitignore files)?

First, the "Searching for .gitignore" message on the status bar stays there longer than the actual .gitignore processing. This is the result of process running on worker threads and only updating the UI occasionally.  If the UI is updated after every background action, you couldn't read it fast enough and it slows down the whole process.

But comparing the total search time with or without .gitignore does give a good indication of the impact on the total processing time. And yes, in some repos using the .gitignore filter slows down the search. It's not just .gitignore, I see the same behavior using exclusion filters entered on the main UI or using [dnGrep.ignore filters](https://github.com/dnGrep/dnGrep/wiki/File-Filters).

I timed generating the exclusion list from .gitignore on several repos, and it has taken at most 1 second, usually around 100 ms.  So, caching the filter list isn't going to help.  

dnGrep is asking git to generate the list using the command `git status --short --ignored`. When using .gitignore, dnGrep always includes the .git directory and the .gitignore file in the exclusion list.

As far as I can see, it's not the time required to generate the exclusion list, it the extra time needed to apply the exclusion list to the directories and files as they are enumerated.

Searching the dnGrep repo with and without .gitignore has no performance impact that I can see.  Using a *.cs pattern, and searching for an uncommon word, I get results in about 0.25 second for searching 515 cs files with or without using .gitignore.  Here's a breakdown:
generate the .gitignore list of 66 directories and 13 files in 114 ms
with .gitignore, apply the *.cs file filter on 1030 files
without .gitignore, apply the *.cs filter on 11719 files (lots of objects in the .git directory).

On the winget-pkgs repo, there are nearly 70,000 directories and nearly 200,000 files outside of the .git directory. This is the repo that took 1 second to generate the exclusion list from .gitignore - and it doesn't exclude anything except the default .git directory and the .gitignore file added by dnGrep. But it still slows down the search for .ps1 files by a few seconds because these filters have to be applied so many times. Searching in *.yaml files in this repo is unbearably show because there are so many of them.

I haven't yet been able to narrow down exactly what is happening in the directory and file filters to see if it can be optimized.

**Comment by csmwww** [comment]
**My solution:** discover and apply "Patterns to exclude" via https://github.com/dnGrep/dnGrep/wiki/File-Filters and use "deploy\\*,target\\\*,node_modules\\\*" abandoning .gtiIngore in my case except for rare needs since the mass of my delays is elsewhere.

_First, let me thank you for taking such a deep interest in my problem, and apologize for delay in my response. As most of us hopefully providing value I am pulled too thin and related to the many hats I wear I have not made time to check my personal email registered in the dnGrep discussion forum for two weeks (pitiful) and just thought to check the forum. Maybe I'll even get to my personal email today, since I'm on vacation!_

**Beyond your hints, how did I get there?** I'd forgotten anyone wrote manuals these days, especially with the GitHub defaults like https://github.com/dnGrep/dnGrep presenting a "source code" perspective to my unhappiness. But now that I get my mind right I find you've done far better than most on the Wiki and https://github.com/dnGrep/dnGrep/wiki/File-Filters set me straight. Here I would only note that in Windows the syntax "node_modules/\*" not working, only "node_modules\*" (unintuitive) and "node_modules\\*", threw this cosmopolitan for a bit of a loop.

**Boring details you deserve having spent so much to try and understand/help me already:**

In trying to determine what's up given your new info, and if there was possible accommodation for my somewhat unique challenge, it was especially helpful to for me to understand dnGrep is not searching .gitIgnore files itself, and is leveraging "git status --short --ignored" which runs about 4 seconds on my "active developer checked out" repo (yes, many .gitIgnore files as well as environment size), and then slightly further that the status message resolution is not as specific to this task as I'd interpreted.

I believe my .gitIgnore concern is death by many cuts, with a relatively large repo containing many related components (projects), as well as many artifacts following a build (vast amounts of which are more source code, for example NodeJS modules downloading tons of *.js files). Search of a clean repo pull is not bad, even with .gitIgnore enabled ("git status --short --ignored" instantaneous under 1sec) though not required in a clean state, but a developer typically runs "built" and search must then be limited in some way to avoid searching build/built artifacts.

With other tools I've limited out vast content by "folder", for example "myproject\build\deploy" containing things like NodeJS\node_modules, but in the breadth and depth of dnGrep I'd not seen such an apparent option - https://github.com/dnGrep/dnGrep/wiki/File-Filters gets my mind right. Patterns to exclude naturally came acress as file name patterns, and the Filter menu being empty baffled me, but I've awoken.

My mistaken approach: not recognizing folder exclusion was available, I went down the path that Git's own similar needs not to be overwhelmed by the massive built self-encapsulated project folder were augmented by Git's natural database of "checked in files" plus .gitIgnore and overloading .gitIgnore with folder identifications (of things Git already ignored via it DB) I'd hoped .gitIngore would achieve exclusion needs. However, enabling .gitIgnore dragged things down, apparently mostly by "git status --short --ignored" in our context taking longer than reasonably expected [and effecting my Git operations as well where I was not as sensitive!] and then perhaps applying the result (and who knows what else).

Very minor note: https://github.com/dnGrep/dnGrep/issues/218 comes up in search, and JVimes references to "Funnel", "Paths" and "ignore" some terms of which did not even appear to apply back in 2015, might be edited for accuracy and clarify to help others who don't make it to the manual - search is still "the way" now until ChatGPT/CodePilot gets a few more cycles better.

Thank you again for so much attention, a great tool, and the depth of work you have done - the later of which I know often makes it hard to present all the configuration options in a way everyone will grasp without sometimes whining like a little baby first!

Regards,
Cris

**Comment by doug24** [comment, marked as answer]
I have some changes to how directory recursion and file filters are applied in the searching for files that may help you both with the file exclusion filters or .gitignore.  But I don't have a repo large enough to see the differences you are seeing.

If you have a directory named "node_modules", you can exclude that directory in the dnGrep syntax using `node_modules\*` (in wildcard) or `node_modules\\.*$` (in regex), using backslashes, not slashes.  As you've seen, you can also use `node_modules` but that could interpreted as either part of a directory name or a file name.  dnGrep is creating a regex from the pattern and testing if the pattern has a match for the current path or filename, using the full path.

Issue #218 is confusing because the image in the issue is linked from the Wiki help pages, and that image has been updated in the Wiki as the layout in the UI has changed.  Here is what the image looked like at the time that issue was created. It was an unusual UI feature where the file filter elements appeared in a combo drop-down box:

---- **obsolete image** ----

![file-filter](https://github.com/user-attachments/assets/bd506738-8287-4e06-94c9-aafb20640fd2)

**Reply by csmwww** [reply]
You're a machine sir; respect - an example to strive for. We've come to the same place (my comments matched yours but I failed to properly escape some wiki markup syntax), and it is kind of you to have looked even deeper and found more optimizations. But, just getting my mind into exclusions as supporting folders name patterns as well as file patterns has me rocking, and happy, and any other improvements will just be gravy for the next rainy day. Thank you!
